rep-2014.rst

```diff
@@ -0,0 +1,243 @@
+
+Value for stakeholders:
+
+- Package maintainers can use these guidelines to integrate performance benchmarking tools (e.g. instrumentation) and data (e.g. results, plots and datasets) in their packages.
+
+- Consumers can use the guidelines in the REP to benchmark ROS Nodes and Graphs in an architecture-neutral, representative, and reproducible manner, as well as the corresponding performance data offered in ROS packages to set expectations on the capabilities of each.
+
+- Hardware vendors and robot manufacturers can use these guidelines to show evidence of the performance of their systems solutions with ROS in an architecture-neutral, representative, and reproducible manner.
+
+The guidelines in here are intended to be a living document that can be updated as new information becomes available.
+
+
+A quantitative approach
+-----------------------
+Performance must be studied with real examples and measurements on real robotic computations, rather than simply as a collection of definitions, designs and/or marketing actions. When creating benchmarks, prefer to use realistic data and situations rather than data designed to test one edge case. Follow the quantitative approach [1]_ when designing your architecture.
+
+
+Approaches to benchmark performance
+-----------------------------------
+There're different types of benchmarking approaches, specially when related to performance. The following definitions clarify the most popular terms inspired by [2]_:
+
+- `Functional performance testing`: Functional performance is the measurable performance of the system’s functions which the user can experience. For example, in a motion planning algorithm, measures could include items like the ratio the algorithm gets the motion plan right.
+  
+- `Non-functional performance testing`: The measurable performance of those aspects that don't belong to the system's functions. In the motion planning example, the latency the planner, the memory consumption, the CPU usage, etc.
+
+
+- `Black-box performance testing`: Measures performance by eliminating the layers above the *layer-of-interest* and replacing those with a specific test application that stimulates the layer-of-interest in the way you need it. This allows to gather the measurement data right inside your specific test application. The acquisition of the data (the probe) resides inside the test application. A major design constraint in such a test is that you have to eliminate the “application” in order to test the system. Otherwise, the real application and the test application would likely interfere.
```

Comment:
I think this name "black box test" is a little bit inaccurate. If the text is correct and only layers above should be eliminated maybe a more accurate name would be integration test. Meaning that the goal of the test is to measure how one layer interact with others or with the test application. Maybe the focus should be moved to the impossibility to access the internal state of the system.

Reply:
See https://drive.google.com/file/d/15nX80RK6aS8abZvQAOnMNUEgh7px9V5S/view for past references to black box tests.